File: broadscope-bailian-sdk-python/broadscope_bailian/api/access_token_client.py
import time
from typing import Tuple

from alibabacloud_bailian20230601.client import Client
from alibabacloud_bailian20230601.models import CreateTokenRequest
from alibabacloud_tea_openapi.models import Config

import broadscope_bailian


class AccessToken:
    def __init__(self, token: str, expired_time: int):
        self.token = token
        self.expired_time = expired_time


class AccessTokenClient:
    def __init__(self, access_key_id: str,
                 access_key_secret: str,
                 agent_key: str = None,
                 endpoint: str = None):
        if endpoint is None or endpoint == "":
            endpoint = broadscope_bailian.pop_endpoint

        self.config = Config(access_key_id=access_key_id,
                             access_key_secret=access_key_secret,
                             endpoint=endpoint)
        self.agent_key = agent_key
        self.access_token = None

    def create_token(self, agent_key) -> Tuple[str, int]:
        """
        create access token for access SFM API

        :param agent_key: agent key of SFM
        :return: token and expired time with unix format
        """

        client = Client(config=self.config)

        token_response = client.create_token(CreateTokenRequest(agent_key=agent_key))
        if token_response.status_code != 200 or token_response.body is None:
            raise RuntimeError("create token error, code=%d" % token_response.status_code)

        token_body = token_response.body
        if not token_body.success:
            request_id = token_body.request_id
            if not request_id:
                request_id = token_response.headers.get("x-acs-request-id")
            raise RuntimeError("create token error, code=%s, message=%s RequestId: %s"
                               % (token_body.code, token_body.message, request_id))

        return token_body.data.token, token_body.data.expired_time

    def get_token(self):
        """ get token if not created or expired """

        timestamp = int(time.time())
        if self.access_token is None or (self.access_token.expired_time - 600) < timestamp:
            token, expired_time = self.create_token(self.agent_key)
            self.access_token = AccessToken(token, expired_time)

        return self.access_token.token

File: broadscope-bailian-sdk-java/src/main/java/com/aliyun/broadscope/bailian/sdk/AccessTokenClient.java
/*
 * All rights Reserved, Designed By Alibaba Group Inc.
 * Copyright: Copyright(C) 1999-2023
 * Company  : Alibaba Group Inc.
 */
package com.aliyun.broadscope.bailian.sdk;

import com.aliyun.bailian20230601.Client;
import com.aliyun.bailian20230601.models.CreateTokenRequest;
import com.aliyun.bailian20230601.models.CreateTokenResponse;
import com.aliyun.bailian20230601.models.CreateTokenResponseBody;
import com.aliyun.broadscope.bailian.sdk.consts.ConfigConsts;
import com.aliyun.broadscope.bailian.sdk.models.AccessToken;
import com.aliyun.teaopenapi.models.Config;
import org.apache.commons.lang3.StringUtils;

/**
 * Title api access token.<br>
 * Description api access token.<br>
 * Created at 2023-07-03 12:49
 *
 * @author yuanci.ytb
 * @version 1.0.0
 * @since jdk8
 */

public class AccessTokenClient {
    private static final int TOKEN_CREATE_THRESHOLD_MINUTES = 10;

    private final String accessKeyId;

    private final String accessKeySecret;

    private final String agentKey;

    private String endpoint = ConfigConsts.POP_ENDPOINT;

    private AccessToken accessToken;

    public AccessTokenClient(String accessKeyId, String accessKeySecret, String agentKey) {
        this.accessKeyId = accessKeyId;
        this.accessKeySecret = accessKeySecret;
        this.agentKey = agentKey;
    }

    public String getToken() {
        long timestamp = System.currentTimeMillis() / 1000;
        if (accessToken == null ||
                (accessToken.getExpiredTime() - TOKEN_CREATE_THRESHOLD_MINUTES * 60) < timestamp) {
            accessToken = createToken();
        }

        return accessToken.getToken();
    }

    public AccessToken createToken() {
        if (StringUtils.isEmpty(this.accessKeyId)) {
            throw new BaiLianSdkException("access key id is required");
        }

        if (StringUtils.isEmpty(this.accessKeySecret)) {
            throw new BaiLianSdkException("access key secret is required");
        }

        if (StringUtils.isEmpty(this.agentKey)) {
            throw new BaiLianSdkException("agent key is required");
        }

        Config config = new Config().setAccessKeyId(accessKeyId).setAccessKeySecret(accessKeySecret);

        if (StringUtils.isNotEmpty(this.endpoint)) {
            config.setEndpoint(endpoint);
        }

        try {
            Client client = new Client(config);
            CreateTokenRequest request = new CreateTokenRequest().setAgentKey(agentKey);
            CreateTokenResponse response = client.createToken(request);
            CreateTokenResponseBody body = response.getBody();
            if (body == null || !body.success) {
                String error = body == null ? "create token error" : body.message;
                throw new BaiLianSdkException(error);
            }

            CreateTokenResponseBody.CreateTokenResponseBodyData data = body.getData();
            if (data == null) {
                throw new BaiLianSdkException("create token error, data is null");
            }

            return new AccessToken(data.getToken(), data.getExpiredTime());
        } catch (BaiLianSdkException e) {
            throw e;
        } catch (Exception e) {
            throw new BaiLianSdkException(e);
        }
    }

    public void setEndpoint(String endpoint) {
        this.endpoint = endpoint;
    }
}
